src/foam/nanos/rope/ROPEAuthorizer.js

```diff
@@ -0,0 +1,224 @@
+/**
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.rope',
+  name: 'ROPEAuthorizer',
+  implements: [ 'foam.nanos.auth.Authorizer' ],
+
+  javaImports: [
+    'foam.core.FObject',
+    'foam.core.X',
+    'foam.dao.ArraySink',
+    'foam.dao.DAO',
+    'foam.nanos.auth.AuthorizationException',
+    'foam.nanos.auth.Authorizer',
+    'foam.nanos.auth.User',
+    'foam.nanos.rope.*',
+    'java.lang.reflect.*',
+    'java.util.ArrayList',
+    'java.util.List',
+    'java.util.Iterator',
+    'java.util.Map',
+    'static foam.mlang.MLang.*'
+  ],
+
+  properties: [
+    {
+      name: 'ropeDAO',
+      class: 'Object',
+      javaType: 'DAO'
+    },
+    {
+      name: 'targetDAOKey',
+      class: 'String',
+      javaType: 'String'
+    }
+  ],
+
+  methods: [
+    {
+      name: 'authorizeByRope',
+      javaType: 'boolean',
+      args: [
+        { name: 'x', javaType: 'X' },
+        { name: 'obj', javaType: 'FObject' },
+        { name: 'targetDAOKey', javaType: 'String' },
+        { name: 'crudKey', javaType: 'String' },
+        { name: 'propertyKey', javaType: 'String' }
+      ],
+      javaCode: `    
+      if ( getRopeDAO() == null ) setRopeDAO((DAO) x.get("ropeDAO"));
+      List<ROPE> ropes = (List<ROPE>) ((ArraySink) ((DAO) getRopeDAO().where(AND(EQ(ROPE.TARGET_DAOKEY, targetDAOKey)))).select(new ArraySink())).getArray();
+
+      for ( ROPE rope : ropes ) {
+        if ( rope.check(x, obj, "", crudKey, propertyKey) ) return true;
+      }
+      return false; 
+      `
+    },
+    {
+      name: 'getPropertiesUpdated',
+      javaType: 'List<String>',
+      args: [
+        { name: 'oldObj', javaType: 'FObject' },
+        { name: 'obj', javaType: 'FObject' }
+      ],
+      documentation: `
+      this function returns a list of the names of properties changed on update or set on create
+      `,
+      javaCode: `
+      try {
+        if ( oldObj == null ) oldObj = (obj.getClass()).newInstance();
+      } catch ( Exception e ) {
+        e.printStackTrace();
+      }
+      Map diff = oldObj.diff(obj);
+      Iterator i = diff.keySet().iterator();
+  
+      List<String> propertyUpdates = new ArrayList<String>();
+      while ( i.hasNext() ) {
+        String propName = (String) i.next();
+        propertyUpdates.add(propName);
+      }
+      return propertyUpdates;
+      `
+    },
+    {
+      name: 'authorizeOnCreate',
+      type: 'Void',
```

Comment:
`type`, `javaThrows`, and `args` are all defined on the interface and don't need to be included here.